src/binary_annotator.h

```diff
@@ -200,17 +206,34 @@ class BinaryAnnotator {
 
   void FixMissingSections();
 
-  template<typename T> inline T GetScalar(uint64_t offset) {
-    return *reinterpret_cast<const T *>(binary_ + offset);
+  template<typename T> inline T GetScalar(const uint64_t offset) const {
+    return ReadScalar<T>(binary_ + offset);
+  }
+
+  inline bool IsValidOffset(const uint64_t offset) const {
+    return offset < binary_length_;
+  }
+
+  // Determines if performing a GetScalar request for `T` at `offset` would read
+  // passed the end of the binary.
+  template<typename T>
+  inline bool IsValidRead(const uint64_t offset,
+                          const uint64_t length = sizeof(T)) {
+    return IsValidOffset(offset + length);
+  }
+
+  void AddSection(const uint64_t offset, const BinarySection &section) {
```

Comment:
Consider leaving a comment saying that the first inserted section wins.

Or you could rename it to something more descriptive, e.g. AddSectionIfNew

Reply:
Done.